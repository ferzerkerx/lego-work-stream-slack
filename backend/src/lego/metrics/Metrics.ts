import { DateMetrics } from './DateMetrics';
import { DateUtils } from '../../utils/DateUtils';

interface CsvDataProvider {
  csvData(): string[][];
}

export class Metrics implements CsvDataProvider {
  categories: string[];
  entries: Array<DateMetrics>;

  constructor(categories: string[], entries: Array<DateMetrics>) {
    this.categories = categories;
    this.entries = entries;
  }

  csvData(): string[][] {
    const data: string[][] = [];

    data.push(['date', ...this.categories]);

    this.entries.forEach((entry: DateMetrics) => {
      let values: string[] = [DateUtils.toPrettyDate(entry.date)];
      for (let category of this.categories) {
        const value = entry.valuesByCategory[category] || 0;
        values.push(`${value}`);
      }
      data.push(values);
    });
    return data;
  }
}
