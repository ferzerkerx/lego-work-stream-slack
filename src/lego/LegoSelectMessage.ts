import {LegoSelectedValue} from "./LegoSelectedValue";

export class LegoSelectMessage {
  id: string;
  selectedValues: LegoSelectedValue[] = [];
  channel: string;
  date: Date = new Date();
}