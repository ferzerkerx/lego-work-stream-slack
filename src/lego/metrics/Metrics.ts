import { DateMetrics } from './DateMetrics';

export class Metrics {
  categories: Array<string>;
  entries: Array<DateMetrics>;
}
