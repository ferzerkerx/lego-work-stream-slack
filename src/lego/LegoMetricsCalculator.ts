import { LegoSelectMessage } from './LegoSelectMessage';
import { LegoSelectedValue } from './LegoSelectedValue';
import { DateUtils } from '../DateUtils';
import { LegoSelectedValueEntry } from './LegoSelectedValueEntry';

export class MetricEntry {
  categories: Array<string>;
  entries: Array<DateEntry>;
}

export class DateEntry {
  date: string;
  values?: any;

  constructor(date: string) {
    this.date = date;
    this.values = {};
  }
}

export class LegoMetricsCalculator {
  static calculate(
    messages: LegoSelectMessage[],
    config: MetricsConfiguration
  ): MetricEntry {
    let datesArray = DateUtils.toDatesArray(
      config.startDate,
      config.endDate,
      config.frequencyInDays
    );

    let categories = new Set<string>();
    let datesEntries = new Map<string, DateEntry>();

    let currentIndex: number = 0;
    for (let message of messages) {
      const dateKey: string = this.keyForMessageDate(
        datesArray,
        currentIndex,
        message.date
      );

      const selectedValues: LegoSelectedValue[] = message.selectedValues;
      let dateEntry: DateEntry =
        datesEntries.get(dateKey) || new DateEntry(dateKey);

      this.updateDateEntry(dateEntry, selectedValues, categories);

      datesEntries.set(dateKey, dateEntry);
    }

    return {
      categories: Array.from(categories),
      entries: Array.from(datesEntries.values()),
    };
  }

  private static updateDateEntry(
    dateEntry: DateEntry,
    selectedValues: LegoSelectedValue[],
    keys
  ): void {
    for (let selectedValue of selectedValues) {
      const sanitizedValueId = LegoMetricsCalculator.sanitizedEntryName(
        selectedValue
      );

      let totalSumForSelectedValueEntry: number =
        dateEntry.values[sanitizedValueId] || 0;

      const currentSumForSelectedValueEntry = LegoMetricsCalculator.sumForSelectedValueEntries(
        selectedValue.entries
      );

      const newTotal: number =
        Number(totalSumForSelectedValueEntry) +
        Number(currentSumForSelectedValueEntry);

      keys.add(sanitizedValueId);
      dateEntry.values[sanitizedValueId] = newTotal;
    }
  }

  private static keyForMessageDate(
    datesArray,
    currentIndex: number,
    messageDate
  ): string {
    let dateInPeriodForMessage: Date = LegoMetricsCalculator.dateInPeriodForMessage(
      datesArray,
      currentIndex,
      messageDate
    );
    return DateUtils.toPrettyDate(dateInPeriodForMessage.toDateString());
  }

  private static dateInPeriodForMessage(
    datesArray: Array<Date>,
    currentIndex: number,
    messageDate
  ): Date {
    let currentPeriod = datesArray[currentIndex];
    while (currentPeriod < messageDate) {
      currentIndex = currentIndex + 1;
      if (currentIndex >= datesArray.length) {
        throw Error('Invalid date range detected');
      }
      currentPeriod = datesArray[currentIndex];
    }
    return currentPeriod;
  }

  private static sumForSelectedValueEntries(
    userEntries: LegoSelectedValueEntry[]
  ): number {
    return userEntries
      .map((entry: LegoSelectedValueEntry) => entry.value)
      .reduce(
        (accumulator: number, currentValue: number) =>
          Number(accumulator) + Number(currentValue)
      );
  }

  private static sanitizedEntryName(selectedValue): string {
    return selectedValue.id.includes('lego-select-option-')
      ? selectedValue.id.substr('lego-select-option-'.length)
      : selectedValue.id;
  }
}
