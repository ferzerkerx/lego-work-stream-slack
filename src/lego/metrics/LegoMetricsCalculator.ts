import { LegoSelectMessage } from '../LegoSelectMessage';
import { LegoSelectedValue } from '../LegoSelectedValue';
import { DateUtils } from '../../DateUtils';
import { LegoSelectedValueEntry } from '../LegoSelectedValueEntry';
import { DateMetrics } from './DateMetrics';
import { Metrics } from './Metrics';
import { CategoryValueMap } from './Types';

interface RawMetrics {
  categories: Set<string>;
  dateEntries: Map<string, DateMetrics>;
}

interface MessageMetrics {
  datePeriodForMessage: string;
  valuesByCategoryForMessage: CategoryValueMap;
}

export class LegoMetricsCalculator {
  static calculate(
    messages: LegoSelectMessage[],
    config: MetricsConfiguration
  ): Metrics {
    const rawMetrics: RawMetrics = this.rawMetrics(config, messages);

    let dateMetrics = rawMetrics.dateEntries;
    if (config.isPercentage) {
      dateMetrics = this.toPercentage(dateMetrics);
    }
    return {
      categories: Array.from(rawMetrics.categories),
      entries: Array.from(dateMetrics.values()),
    };
  }

  private static toPercentage(
    dateMetrics: Map<string, DateMetrics>
  ): Map<string, DateMetrics> {
    for (let dateEntry of dateMetrics.values()) {
      const categoryValueMapForDateEntry: CategoryValueMap =
        dateEntry.valuesByCategory;

      const totalValueForDateEntry = this.totalValueForDateEntry(
        categoryValueMapForDateEntry
      );

      if (totalValueForDateEntry > 0) {
        for (let category of Object.keys(categoryValueMapForDateEntry)) {
          const percentageValue =
            (categoryValueMapForDateEntry[category] / totalValueForDateEntry) *
            100;
          categoryValueMapForDateEntry[category] = Math.round(percentageValue);
        }
      }
    }
    return dateMetrics;
  }

  private static totalValueForDateEntry(
    categoryValueMapForDateEntry: CategoryValueMap
  ): number {
    return Object.keys(categoryValueMapForDateEntry)
      .map(category => categoryValueMapForDateEntry[category])
      .reduce(
        (accumulator: number, currentValue: number) =>
          Number(accumulator) + Number(currentValue)
      );
  }

  private static rawMetrics(
    config: MetricsConfiguration,
    messages: LegoSelectMessage[]
  ): RawMetrics {
    const datePeriods: Date[] = DateUtils.toDatesArray(
      config.startDate,
      config.endDate,
      config.frequencyInDays
    );

    const categories = new Set<string>();
    const dateEntries = new Map<string, DateMetrics>();

    for (let message of messages) {
      const metricsForMessage: MessageMetrics = this.metricsForMessage(
        message,
        datePeriods
      );

      const dateMetrics: DateMetrics =
        dateEntries.get(metricsForMessage.datePeriodForMessage) ||
        new DateMetrics(metricsForMessage.datePeriodForMessage);
      const mergedDateMetrics = dateMetrics.merge(
        metricsForMessage.valuesByCategoryForMessage
      );

      Object.keys(metricsForMessage.valuesByCategoryForMessage).forEach(
        category => categories.add(category)
      );

      dateEntries.set(
        metricsForMessage.datePeriodForMessage,
        mergedDateMetrics
      );
    }
    return {
      categories: categories,
      dateEntries: dateEntries,
    };
  }

  private static metricsForMessage(
    message,
    datePeriods: Date[]
  ): MessageMetrics {
    const selectedValues: LegoSelectedValue[] = message.selectedValues;
    const valuesByCategoryForMessage = this.valuesByCategory(selectedValues);

    const datePeriodForMessage: string = this.datePeriodForMessage(
      datePeriods,
      message.date
    );

    return {
      datePeriodForMessage: datePeriodForMessage,
      valuesByCategoryForMessage: valuesByCategoryForMessage,
    };
  }

  private static valuesByCategory(
    selectedValues: LegoSelectedValue[]
  ): CategoryValueMap {
    let categoryValueMap: CategoryValueMap = {};
    for (let selectedValue of selectedValues) {
      const categoryName = LegoMetricsCalculator.extractCategoryName(
        selectedValue
      );

      let totalSumForCategory: number = categoryValueMap[categoryName] || 0;

      const currentSumForCategory = LegoMetricsCalculator.extractSumForCategory(
        selectedValue.entries
      );

      categoryValueMap[categoryName] =
        Number(totalSumForCategory) + Number(currentSumForCategory);
    }
    return categoryValueMap;
  }

  private static datePeriodForMessage(datesArray, messageDate): string {
    let dateInPeriodForMessage: Date = LegoMetricsCalculator.dateInPeriodForMessage(
      datesArray,
      messageDate
    );
    return DateUtils.toPrettyDate(dateInPeriodForMessage.toDateString());
  }

  private static dateInPeriodForMessage(
    datePeriods: Array<Date>,
    messageDate
  ): Date {
    let currentIndex: number = 0;
    let currentPeriod = datePeriods[currentIndex];
    while (currentPeriod < messageDate) {
      currentIndex = currentIndex + 1;
      if (currentIndex >= datePeriods.length) {
        throw Error('Invalid date range detected');
      }
      currentPeriod = datePeriods[currentIndex];
    }
    return currentPeriod;
  }

  private static extractSumForCategory(
    selectedValueEntries: LegoSelectedValueEntry[]
  ): number {
    return selectedValueEntries
      .map(selectedValueEntry => selectedValueEntry.value)
      .reduce(
        (accumulator: number, currentValue: number) =>
          Number(accumulator) + Number(currentValue)
      );
  }

  private static extractCategoryName(selectedValue): string {
    return selectedValue.id.includes('lego-select-option-')
      ? selectedValue.id.substr('lego-select-option-'.length)
      : selectedValue.id;
  }
}
