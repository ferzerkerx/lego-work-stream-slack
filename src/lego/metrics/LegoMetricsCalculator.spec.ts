import { LegoSelectMessage } from '../LegoSelectMessage';
import { LegoMetricsCalculator } from './LegoMetricsCalculator';
import { Metrics } from './Metrics';

function givenLegoMessages(): LegoSelectMessage[] {
  return [
    {
      id: 'someId',
      date: new Date('2018-10-10'),
      selectedValues: [
        {
          id: 'someKey',
          entries: [
            { userData: { userId: 'userOne', userName: 'userOne' }, value: 10 },
            { userData: { userId: 'userTwo', userName: 'userTwo' }, value: 5 },
          ],
        },
      ],
      channelData: { channelId: 'someChannel', name: 'name' },
    },
    {
      id: 'someCoolId',
      date: new Date('2018-10-10'),
      selectedValues: [
        {
          id: 'someKey',
          entries: [
            { userData: { userId: 'userOne', userName: 'userOne' }, value: 1 },
          ],
        },
        {
          id: 'someOtherKey',
          entries: [
            { userData: { userId: 'userOne', userName: 'userOne' }, value: 6 },
            {
              userData: { userId: 'userThree', userName: 'userThree' },
              value: 3,
            },
          ],
        },
      ],
      channelData: { channelId: 'someChannel', name: 'name' },
    },
    {
      id: 'someOtherId',
      date: new Date('2018-10-11'),
      selectedValues: [
        {
          id: 'someKey',
          entries: [
            { userData: { userId: 'userOne', userName: 'userOne' }, value: 2 },
            { userData: { userId: 'userTwo', userName: 'userTwo' }, value: 1 },
          ],
        },
        {
          id: 'someOtherKey',
          entries: [
            { userData: { userId: 'userOne', userName: 'userOne' }, value: 6 },
            {
              userData: { userId: 'userThree', userName: 'userThree' },
              value: 7,
            },
          ],
        },
      ],
      channelData: { channelId: 'someChannel', name: 'name' },
    },
  ];
}

describe('LegoMetricsCalculator', () => {
  test('calculate', () => {
    let config: MetricsConfiguration = {
      startDate: new Date('2018-10-10'),
      endDate: new Date('2018-10-11'),
      frequencyInDays: 1,
    };

    const metrics: Metrics = LegoMetricsCalculator.calculate(
      givenLegoMessages(),
      config
    );
    expect(metrics).toMatchSnapshot();
  });

  test('calculate with periods', () => {
    let config: MetricsConfiguration = {
      startDate: new Date('2018-10-08'),
      endDate: new Date('2018-10-12'),
      frequencyInDays: 3,
    };

    const metrics: Metrics = LegoMetricsCalculator.calculate(
      givenLegoMessages(),
      config
    );
    expect(metrics).toMatchSnapshot();
  });

  test('calculate with periods and percentage', () => {
    let config: MetricsConfiguration = {
      startDate: new Date('2018-10-08'),
      endDate: new Date('2018-10-12'),
      frequencyInDays: 3,
      isPercentage: true,
    };

    const metrics: Metrics = LegoMetricsCalculator.calculate(
      givenLegoMessages(),
      config
    );
    expect(metrics).toMatchSnapshot();
  });

  test('metrics csv data', () => {
    let config: MetricsConfiguration = {
      startDate: new Date('2018-10-10'),
      endDate: new Date('2018-10-11'),
      frequencyInDays: 1,
    };

    const metrics: Metrics = LegoMetricsCalculator.calculate(
      givenLegoMessages(),
      config
    );
    expect(metrics.csvData()).toMatchSnapshot();
  });
});
