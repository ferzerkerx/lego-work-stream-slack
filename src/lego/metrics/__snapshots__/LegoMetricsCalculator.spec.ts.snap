// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LegoMetricsCalculator calculate 1`] = `
Object {
  "categories": Array [
    "someKey",
    "someOtherKey",
  ],
  "entries": Array [
    DateMetrics {
      "date": "2018-10-10",
      "valuesByCategory": Object {
        "someKey": 16,
        "someOtherKey": 9,
      },
    },
    DateMetrics {
      "date": "2018-10-11",
      "valuesByCategory": Object {
        "someKey": 3,
        "someOtherKey": 13,
      },
    },
  ],
}
`;

exports[`LegoMetricsCalculator calculate with periods 1`] = `
Object {
  "categories": Array [
    "someKey",
    "someOtherKey",
  ],
  "entries": Array [
    DateMetrics {
      "date": "2018-10-11",
      "valuesByCategory": Object {
        "someKey": 19,
        "someOtherKey": 22,
      },
    },
  ],
}
`;

exports[`LegoMetricsCalculator calculate with periods and percentage 1`] = `
Object {
  "categories": Array [
    "someKey",
    "someOtherKey",
  ],
  "entries": Array [
    DateMetrics {
      "date": "2018-10-11",
      "valuesByCategory": Object {
        "someKey": 46,
        "someOtherKey": 54,
      },
    },
  ],
}
`;
